src/Language/AST/NodeList.php

```diff
@@ -6,101 +6,127 @@
 
 use ArrayAccess;
 use Countable;
-use Generator;
 use GraphQL\Utils\AST;
+use InvalidArgumentException;
 use IteratorAggregate;
+use Traversable;
 use function array_merge;
 use function array_splice;
 use function count;
 use function is_array;
 
+/**
+ * @template T of Node
+ * @phpstan-implements ArrayAccess<int|string, T>
+ * @phpstan-implements IteratorAggregate<T>
+ */
 class NodeList implements ArrayAccess, IteratorAggregate, Countable
 {
-    /** @var Node[]|mixed[] */
+    /**
+     * @var Node[]
+     * @phpstan-var array<T>
+     */
     private $nodes;
 
     /**
-     * @param Node[]|mixed[] $nodes
+     * @param Node[] $nodes
      *
-     * @return static
+     * @phpstan-param array<T> $nodes
+     * @phpstan-return self<T>
      */
-    public static function create(array $nodes)
+    public static function create(array $nodes) : self
     {
-        return new static($nodes);
+        return new self($nodes);
     }
 
     /**
-     * @param Node[]|mixed[] $nodes
+     * @param Node[] $nodes
+     *
+     * @phpstan-param array<T> $nodes
      */
     public function __construct(array $nodes)
     {
         $this->nodes = $nodes;
     }
 
     /**
-     * @param mixed $offset
-     *
-     * @return bool
+     * @param int|string $offset
      */
-    public function offsetExists($offset)
+    public function offsetExists($offset) : bool
     {
         return isset($this->nodes[$offset]);
     }
 
     /**
-     * @param mixed $offset
+     * TODO this *should* always return a Node, but sometimes a string is in here?
```

Comment:
@vladar can you enlighten us on what is going on here?

During some tests, the `Printer` puts items such as `"input:SomeInput"` in a NodeList and mutates it somehow? That makes static validation quite hard/impossible, since we can no longer guarantee that the type `T` is always the same.

Reply:
Printer uses a Visitor which can transform any node to an arbitrary data structure (in case of printer it transforms a node to a printed version). See here:

https://github.com/webonyx/graphql-php/blob/34fe9615097be1f1d7f628ef563489973905e759/src/Language/Visitor.php#L31-L58

Technically visitor doesn't mutate but creates a new structure (clones where appropriate). I guess for strict typing we will have to adjust Visitor and replace NodeList with array (or less strict List structure when modifications are made).

Reply:
Switching to an array when transforming seems fine, since i think the lazyness aspect of `NodeList` does not matter at this point anymore.

How about we open a separate PR for that and merge this one as is to enable further work on phpstan:0.12 compatibility?

Reply:
Makes sense, we have a TODO in place there so can revisit when Visitor part is unblocked